the-collab-lab/tcl-26-smart-shopping-list
# Uncheck feature

## Description

This PR adds functionality to handle the user checking an item that is already checked. Previously, checkboxes for items marked as purchased were just disabled. This update allows us to have active checkboxes at all times, which has the benefit of keeping the focus state as a checkbox is checked. Finally, additional notice/error messages are provided to the user when checking/unchecking items. These are set in the new `itemNotice` state, which is an object with properties related to the type and content of the message (including whether it should use aria assertive or polite, whether it should be styled as an error, whether it should be only for screen readers so visually hidden.)

### Unchecking to undo an accidental purchase
When a user checks an item as purchased, there is now a 5 minute window during which they can uncheck the item, allowing them to correct an accidental purchase. To ensure the information doesn't get destroyed when an item is checked off and can therefore be restored, the `checkAsPurchased` function now adds an extra field called `undoRestore` used to save a copy of the previous lastPurchaseDate and purchaseInterval.  

When an item is unchecked, `handleUncheck` runs a function called `isPurchaseWithinUndoWindow` to verify if it's still within the 5 minute "undo window." If it is, the `uncheckAsPurchased` function passed down from `ShoppingList` handles restoring the old data in the Firestore entry for the item. A hidden user noticefor screen readers is then printed saying the purchase was undone.

### Unchecking outside the undo window
If more than 5 minutes has passed, the user is not allowed to uncheck an item, since it's not likely to be accidental and would affect their purchase history tracking. In this case, a user notice is printed saying "You already purchased this in the last 24 hours" so that the user can understand why their action failed. This message appears inline underneath the item in the list to keep things uncluttered and simple to understand, and is destroyed once the checkbox loses focus. This method was chosen over a time out to give the user more control over the length of time the message is shown, with accessibility in mind. 

### New user notices when checking an item as purchased

Since a system for showing item-related messages is now in place, this was a good opportunity to add two extra user feedback messages in the same way. When a user marks an item as purchased, a screen reader only message is printed (ie. "Purchased bananas". If something fails with this operation, a "Sorry, there was a problem" error is printed and available to the user both visually and as a screen reader alert. 

## Related Issue

Closes #40 

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

![image](https://user-images.githubusercontent.com/67769490/131171927-818e47e3-3913-4a7e-a71d-1675d3225d02.png)

![image](https://user-images.githubusercontent.com/67769490/131172895-9980aef2-dac8-44c1-a3aa-e6f8788cb901.png)

## Testing Steps / QA Criteria

1. Pull down the branch and install dependencies if you have not already.
2. Run npm start to start the local server.
3. Try unchecking an item that is already checked. You should see a notice underneath the item.
4. Try checking an item, and then unchecking. You should be able to uncheck if it's within 5 minutes of checking. Be sure to test with multiple items, including new items with no lastPurchaseDate, and items that have been purchased a few times.
5. If you have a screen reader, you can try checking and unchecking items, and listening for the visually hidden user notices.
